po/masks arrows - better drawing

TurboGit opened the pull request:
A bit of love for the masks arrows to/from the source & destination.

Before:
![image](https://user-images.githubusercontent.com/467069/229307095-da907076-fead-4eea-9c3e-1f76ac0a0ed2.png)

200%
![image](https://user-images.githubusercontent.com/467069/229307132-41e1541d-d827-4819-a8f2-dc63b8caef5f.png)

400% (tiny arrow):
![image](https://user-images.githubusercontent.com/467069/229307172-d4d0b8f1-5a00-4817-96a0-05342c5c8348.png)

After:
![image](https://user-images.githubusercontent.com/467069/229307297-7ccc66fe-ac0f-49cf-99b4-bb45a59d90e6.png)

200%:
![image](https://user-images.githubusercontent.com/467069/229307273-553040b9-21a5-4e98-9f9a-3c7a6affc622.png)

400% (tiny arrow):
![image](https://user-images.githubusercontent.com/467069/229307243-14e22da8-bdc8-4991-82b9-a52300d007a1.png)

elstoc commented:
IMO the arrow is a little large when fully zoomed out. I think it does make sense for the arrow to get larger as you zoom (as in current master) and it may be an optical illusion but the arrow head feels larger when zoomed out than when zoomed in, which feels wrong. Perhaps some compromise can be reached between master (which is a bit too small when zoomed out) and this PR (which is a bit too large).

It also feels like the arrow head is a bit too far away from the actual shape when zoomed out. In master it seems to touch the inner edge of the feathering. In this PR it seems to touch the outer edge (at least when zoomed out) but of course unless your mouse is on it, the outer edge is invisible, and the gap is noticeable.

TurboGit commented:
@elstoc : Thanks for testing.

> I think it does make sense for the arrow to get larger as you zoom

Should be the case now.



> Edit: the distance is even more noticeable with a path:

A bug in the path taking the control points into the calculation of the minimal distance. Now fixed.

> It also feels like the arrow head is a bit too far away from the actual shape when zoomed out.

Adjusted, should also be better now.

elstoc commented:
Thanks those look a bit better now. A couple of minor niggles (not sure if they're the same in current master or not but while we're here...):

For circles, the line goes from the middle of the source to the edge of the target, whereas for paths, it's always edge-to-edge. Feels like this could be more consistent.

![Screenshot_2023-04-02_18-09-58](https://user-images.githubusercontent.com/9555491/229368045-638c9dd9-9ca9-457e-b926-20f29530f6db.png)

Also there are some corner cases where there is no line, only an arrow but there does still seem to be some logic to shift the line to point to different edges. For example

![Screenshot_2023-04-02_18-12-05](https://user-images.githubusercontent.com/9555491/229368154-034adac6-0045-4be1-99ff-54bb18a48f98.png)

Could maybe some minimum line length be used to guard against this? Obviously there will be cases where source and target are almost on top of one another but it feels like this could probably be improved.

TurboGit commented:
> For circles, the line goes from the middle of the source to the edge of the target, whereas for paths, it's always edge-to-edge.

Agreed, circle & ellipse fixed.

> Could maybe some minimum line length be used to guard against this?

I've put that in place, but not very good. Probably ok for the PR and what we could try is when the arrow is too small start from the center of the source & target area. It will not be clean either when both area are very close, but I suppose this is a corner case and not really something that should happen.

elstoc commented:
Hmm now there are places where we lose the arrow entirely, which IMO is worse than just seeing the arrow head (which at least allowed you to see shape was which). What I meant was that if the arrow line is too small perhaps it should switch to pointing to the far edge. It does this already but a bit later. I mean, for example, flip from this...

![Screenshot_2023-04-02_19-05-09](https://user-images.githubusercontent.com/9555491/229370722-3178a2ec-2e56-473f-9f7a-1b34d309f0b1.png)

To this...

![Screenshot_2023-04-02_19-05-21](https://user-images.githubusercontent.com/9555491/229370730-34361710-677d-4388-8bbd-db71c1d534f4.png)

But without first showing just the arrow heads or no arrow at all i.e. when the arrow gets below a particular size, try flipping to another edge

TurboGit commented:
> What I meant was that if the arrow line is too small perhaps it should switch to pointing to the far edge.

It will cross the shape, not really good. I have a new implementation where the arrow is not displayed only the line linking both shaped. I suppose this is corner case as I have never used overlapping shapes.

TurboGit commented:
> i.e. when the arrow gets below a particular size, try flipping to another edge

I have no idea how to do that. Another edge, but which one? On screen you see this, but I don't have the math to have a clear model for this.

elstoc commented:
Ok I prefer it as it was before I started complaining about it :)

Arrow head without a line is better than line without an arrow head.

TurboGit commented:
Reverted to have the arrow. We'll see if we can find a better way in some future work.

TurboGit commented:
Just added a small refinement of arrow display for path. For complex form the arrow was not always the smallest between the source & dest.

elstoc commented:
There are still some cases where the circle mask draws its line from the centre of the shape, but again, fairly corner-case. For example

![Screenshot_2023-04-02_19-47-46](https://user-images.githubusercontent.com/9555491/229372676-221b0f90-c78d-4e6c-8c53-d0a0e4c0ac5f.png)

TurboGit commented:
> There are still some cases where the circle mask draws its line from the centre of the shape, but again, fairly corner-case. For example

Should be fixed, the circle control point (center) was not ignore when computing the closest point for the source.

elstoc commented:
Yeah the whole algorithm for the position of the connecting arrow needs some work but I agree it's not an easy problem and not one for this PR. Looks good to me now.

AlicVB commented on src/develop/masks/circle.c:
```diff
@@ -942,21 +941,25 @@ static void _circle_events_post_expose(cairo_t *cr,
+      // (to_x, to_y) is the point of intersection
+      const float to_x = gpt->points[0] + radius * cosf(cangle);
+      const float to_y = gpt->points[1] + radius * sinf(cangle);
+
```
note : 
- (a very old omission I just spot) that only works without any distortion, we may want to apply distortion to the resulting point... I've not checked, but chance are that this also apply to the other shapes !
- same computation can also be done with source point. That may gain some cpu cyles...

TurboGit replied:
> * (a very old omission I just spot) that only works without any distortion, we may want to apply distortion to the resulting point... I've not checked, but chance are that this also apply to the other shapes !

Fixed. And no this is not the case for other shapes which are using the distorted set of points.

TurboGit commented:
Some more code refactoring/sharing. And makes brush arrow consistent with what is done for other shapes.

This makes better display, for example, before:
![image](https://user-images.githubusercontent.com/467069/229431695-d7009972-0a3d-4b30-9a74-4bba4dde16a7.png)

After:
![image](https://user-images.githubusercontent.com/467069/229431447-abcf4e32-5538-4d3d-8740-801bb2843876.png)
